File: php/src/Domain/ModifierValidation.php
<?php

declare(strict_types=1);

namespace App\Domain;

/**
 * Validacion de selecciones de modificadores contra su grupo.
 *
 * Pura y testeable: recibe la configuracion ya resuelta (min_select/max_select
 * por grupo) y cuantos modificadores de ese grupo trae la linea del pedido.
 */
final class ModifierValidation
{
    public static function validateSelection(ModifierGroupConstraint $constraint, int $selectedCount): void
    {
        if ($constraint->isRequired && $selectedCount === 0) {
            throw new ModifierValidationError("El grupo '{$constraint->name}' es obligatorio");
        }
        if ($selectedCount < $constraint->minSelect) {
            throw new ModifierValidationError(
                "El grupo '{$constraint->name}' requiere minimo {$constraint->minSelect} seleccion(es)"
            );
        }
        if ($selectedCount > $constraint->maxSelect) {
            throw new ModifierValidationError(
                "El grupo '{$constraint->name}' permite maximo {$constraint->maxSelect} seleccion(es)"
            );
        }
    }
}

File: app/domain/modifier_validation.py
"""Validacion de selecciones de modificadores contra su grupo.

Pura y testeable: recibe la configuracion ya resuelta (min_select/max_select
por grupo) y cuantos modificadores de ese grupo trae la linea del pedido.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class ModifierGroupConstraint:
    group_id: str
    name: str
    min_select: int
    max_select: int
    is_required: bool


class ModifierValidationError(Exception):
    pass


def validate_selection(constraint: ModifierGroupConstraint, selected_count: int) -> None:
    if constraint.is_required and selected_count == 0:
        raise ModifierValidationError(f"El grupo '{constraint.name}' es obligatorio")
    if selected_count < constraint.min_select:
        raise ModifierValidationError(
            f"El grupo '{constraint.name}' requiere mínimo {constraint.min_select} selección(es)"
        )
    if selected_count > constraint.max_select:
        raise ModifierValidationError(
            f"El grupo '{constraint.name}' permite máximo {constraint.max_select} selección(es)"
        )
